### DISCUSSION TITLE
Formulating the message to be sent to each agent via the phased coordinator

### DISCUSSION THREAD

**Discussion by fjuma** [wise-agents/wise-agents#222] [Ideas]
Currently, as mentioned [here](https://github.com/wise-agents/wise-agents/pull/207#discussion_r1735777198), the phased coordinator implementation takes the original query it receives and simply passes it to the agents for each phase. Later, if the coordinator decides to cycle through the phases, it attempts to rephrase the original query, and then simply passes the rephrased query to all of the agents. This latter part will also require discussion and is tracked via https://github.com/wise-agents/wise-agents/issues/215 but let's focus on just the first part for this discussion.

Instead of just passing the original query as is to each agent, it would be better for the coordinator to pass a more specific query to each agent instead.

One option is for the coordinator to use its LLM to try to formulate the query to be sent to an agent by making use of the following information:

* the name of the phase
* the description of the agent
* the original query

This would be a good starting point for the first phase.

Then for subsequent phases, the query would be formulated taking into account the above information and would also consider the conversation history.

WDYT?

**Comment by maeste** [comment]
> Currently, as mentioned [here](https://github.com/wise-agents/wise-agents/pull/207#discussion_r1735777198), the phased coordinator implementation

This is unrelated to this discussion, but there is something I just noticed looking in the code to answer this thread: The phased coordinator class is named just  Coordinator, while we have a SequentialCoordinator. Maybe we can rename it for clarity the same way, especially if at some point we would need an abstract Coordinator (see #223) 

> takes the original query it receives and simply passes it to the agents for each phase. Later, if the coordinator decides to cycle through the phases, it attempts to rephrase the original query, and then simply passes the rephrased query to all of the agents. This latter part will also require discussion and is tracked via #215 but let's focus on just the first part for this discussion.
> 
> Instead of just passing the original query as is to each agent, it would be better for the coordinator to pass a more specific query to each agent instead.
> 
> One option is for the coordinator to use its LLM to try to formulate the query to be sent to an agent by making use of the following information:
> 
> * the name of the phase
> * the description of the agent
> * the original query
> 
> This would be a good starting point for the first phase.
> 
> Then for subsequent phases, the query would be formulated taking into account the above information and would also consider the conversation history.
> 
> WDYT?

I think the approach is fine. For subsequent phases we can also consider if we are passing the full conversation history (or even using it from the context as is), or maybe we can try to select which content to pass or to summarize for a specific agent. It could be important if we are using in any agent a small model supporting short token length. 
That said, the more I think about this coordinator, the more I feel it needs specific implementation in different use cases (take this summarizing need for small models: it could be needed or not depends not on which agents will be used in each phase and more precisely which modes those agents use). So, considering all these different needs, we may need to design it as generic, configurable and pluggable. We can start a specific discussion about this, and it will also involve the agent self-description, both in NL and more formal definitions, such as the model or agent limit in terms of token history supported.
If it makes sense, please open an issue to track this. After more thought and discussion, we can define a subtask.

There is another aspect of this phase coordinator, which is crucial and I've tried to track in issue #224 . The coordinator, in a pure distributed agents framework architecture, could easily become a single point of failure. Making it clustered could be a long-term solution, but the minimal requirement for it would be being resilient to failure or crash and able to continue the work after a crash and a restart. For this reason, I think all the information about phases/thresholds and all the variables currently at a class level should be moved into context.

**Reply by fjuma** [reply]
> > Currently, as mentioned [here](https://github.com/wise-agents/wise-agents/pull/207#discussion_r1735777198), the phased coordinator implementation
> 
> This is unrelated to this discussion, but there is something I just noticed looking in the code to answer this thread: The phased coordinator class is named just Coordinator, while we have a SequentialCoordinator. Maybe we can rename it for clarity the same way, especially if at some point we would need an abstract Coordinator (see #223)

Yes, this was bothering me too. I've created https://github.com/wise-agents/wise-agents/issues/225 to track renaming it.

> 
> > takes the original query it receives and simply passes it to the agents for each phase. Later, if the coordinator decides to cycle through the phases, it attempts to rephrase the original query, and then simply passes the rephrased query to all of the agents. This latter part will also require discussion and is tracked via #215 but let's focus on just the first part for this discussion.
> > Instead of just passing the original query as is to each agent, it would be better for the coordinator to pass a more specific query to each agent instead.
> > One option is for the coordinator to use its LLM to try to formulate the query to be sent to an agent by making use of the following information:
> > 
> > * the name of the phase
> > * the description of the agent
> > * the original query
> > 
> > This would be a good starting point for the first phase.
> > Then for subsequent phases, the query would be formulated taking into account the above information and would also consider the conversation history.
> > WDYT?
> 
> I think the approach is fine. For subsequent phases we can also consider if we are passing the full conversation history (or even using it from the context as is), or maybe we can try to select which content to pass or to summarize for a specific agent. It could be important if we are using in any agent a small model supporting short token length. That said, the more I think about this coordinator, the more I feel it needs specific implementation in different use cases (take this summarizing need for small models: it could be needed or not depends not on which agents will be used in each phase and more precisely which modes those agents use). So, considering all these different needs, we may need to design it as generic, configurable and pluggable. We can start a specific discussion about this, and it will also involve the agent self-description, both in NL and more formal definitions, such as the model or agent limit in terms of token history supported. If it makes sense, please open an issue to track this. After more thought and discussion, we can define a subtask.

Right, designing this in a generic, configurable way would be good. Even though we know what model an agent's LLM is using, the model name on its own doesn't always indicate the context length and we wouldn't be able to rely on some hardcoded mapping of model names to lengths since new models are always coming out. So having something in the agent configuration to indicate this would be good.

> 
> There is another aspect of this phase coordinator, which is crucial and I've tried to track in issue #224 . The coordinator, in a pure distributed agents framework architecture, could easily become a single point of failure. Making it clustered could be a long-term solution, but the minimal requirement for it would be being resilient to failure or crash and able to continue the work after a crash and a restart. For this reason, I think all the information about phases/thresholds and all the variables currently at a class level should be moved into context.

Yes that's a great point, will move the info that we currently have at a class level over to the context.

**Reply by fjuma** [reply]
> > There is another aspect of this phase coordinator, which is crucial and I've tried to track in issue #224 . The coordinator, in a pure distributed agents framework architecture, could easily become a single point of failure. Making it clustered could be a long-term solution, but the minimal requirement for it would be being resilient to failure or crash and able to continue the work after a crash and a restart. For this reason, I think all the information about phases/thresholds and all the variables currently at a class level should be moved into context.
> 
> Yes that's a great point, will move the info that we currently have at a class level over to the context.

Realized that we actually have all of the state related information as part of the context already.

The `phases`, `max_iterations`, `confidence_score`, and optionally the `system_message` to use are part of the coordinator agent configuration and aren't related to the state itself. Note that `phases` only contains the phase names. If a coordinator agent were to go down and be restarted, it would be restarted with these values.

When a coordinator receives a request, it creates a new chat id for the request and adds the following information to the context to keep track of the state. This information from the context gets updated by the coordinator when processing a response.

* `_agent_phase_assignments`: This maps a chat id to a list that contains a list of agent names to be invoked in each phase
* ` _current_phase`: This maps a chat id to the current phase, represented as an index
* `_required_agents_for_current_phase`: This maps a chat id to a list of agent names that need to be executed for the current phase
* `_queries`: This maps a chat id to a list of queries representing the queries attempted in each iteration (this will need modifications as we come up with a strategy for formulating queries specific to each agent)

EDIT: One thing does need to be moved to the context: `route_response_to`.

I'll add this information to `WiseAgentContext` as part of https://github.com/wise-agents/wise-agents/issues/211.

I'll also update https://github.com/wise-agents/wise-agents/issues/224 to mention this and indicate that we should test the case where the coordinator agent is stopped and restarted to check that we can handle this case.

**Comment by maeste** [comment]
> That said, the more I think about this coordinator, the more I feel it needs specific implementation in different use cases (take this summarizing need for small models: it could be needed or not depends not on which agents will be used in each phase and more precisely which modes those agents use). So, considering all these different needs, we may need to design it as generic, configurable and pluggable. We can start a specific discussion about this, and it will also involve the agent self-description, both in NL and more formal definitions, such as the model or agent limit in terms of token history supported.
> If it makes sense, please open an issue to track this. After more thought and discussion, we can define a subtask.

Thinking more about this, it probably can be achieved composing it with more agents. So for example we can have an agent A responsible to extract info relevant for agent B from the context. Of course A need to be called before B in a sequence, which introduce more complexity. But of course is hard to don't have more complexity if we want more flexibility. 
Most likely the best solution could be mixing the 2 approaches I'm mentioning. 
I feel we need a discussion or design session to discuss all these aspects

Best regards
S.

**Reply by fjuma** [reply]
> > That said, the more I think about this coordinator, the more I feel it needs specific implementation in different use cases (take this summarizing need for small models: it could be needed or not depends not on which agents will be used in each phase and more precisely which modes those agents use). So, considering all these different needs, we may need to design it as generic, configurable and pluggable. We can start a specific discussion about this, and it will also involve the agent self-description, both in NL and more formal definitions, such as the model or agent limit in terms of token history supported.
> > If it makes sense, please open an issue to track this. After more thought and discussion, we can define a subtask.
> 
> Thinking more about this, it probably can be achieved composing it with more agents. So for example we can have an agent A responsible to extract info relevant for agent B from the context. Of course A need to be called before B in a sequence, which introduce more complexity. But of course is hard to don't have more complexity if we want more flexibility. Most likely the best solution could be mixing the 2 approaches I'm mentioning. I feel we need a discussion or design session to discuss all these aspects

To make sure I'm understanding correctly, is agent A in your example an agent that is only used for extracting relevant info from the context for another agent? So this agent A would be called to extract the relevant info for any agent that requires a shortened context?
